### PR TITLE
Server-side builds (version 0.1.0)

### LINKED ISSUE #413: Generate test database vs. HTML files
To aid in tests, we should do the following (as suggested in https://github.com/foolip/mdn-bcd-collector/issues/376#issuecomment-685021436):

- Generate a database (JSON file) of tests per BCD data path.
- Let the server use that database to generate the test pages, keeping the output of those pages as trivial as possible. The browser should not need to query and interpret MANIFEST.json to run tests.

### LINKED ISSUE #414: Move backend endpoints?
The `/api/` URL contains two separate pieces of functionality in it -- the API tests, and the API endpoints.  We should rename the API endpoint URLs to something else (for example, `/backend/` or just leave them as-is).  Either that, or move the tests themselves to a new URL, which we could do considering #413.

### LINKED ISSUE #417: Stylize pages
I was thinking that we should give the homepage, reports page, and individual test completion pages a little bit of styling to make them look more attractive.  What do you think, @foolip?

### LINKED ISSUE #418: Console API doesn't post updates correctly
In BCD, the Console API is referred to as `api.Console` -- however, the WebIDL actually defines `console` as a namespace with no capitalization.  Since the identifiers are case-sensitive, this causes no updated results.

There's a few solutions I can think of:
- Rename the identifier in the build script -- hacky and special-casing, but it'll do the trick at least
- Rename the identifier in update-bcd -- also hacky and special-casing
- Make identifiers case-insensitive -- could cause unforseen issues in BCD

### LINKED ISSUE #420: Generate endpoints on the fly in backend
> Do we need manifest.json separately from test.json, shouldn't the backend be able to derive everything it needs from just the list of tests?

_Originally posted by @foolip in https://github.com/foolip/mdn-bcd-collector/pull/416#discussion_r484376238_

Additional comments:
https://github.com/foolip/mdn-bcd-collector/pull/416#discussion_r484440344
https://github.com/foolip/mdn-bcd-collector/pull/416#discussion_r484450944

### LINKED ISSUE #437: Use <datalist> for individual test pages
As a follow-up to https://github.com/foolip/mdn-bcd-collector/issues/376, it would be nice to use an `<input list>` with `<datalist>` instead of a `<select>`, as this makes it easier to find the right endpoint by typing part of an API name.

### PR DESCRIPTION
**Pull request by vinyldarkscratch**
This PR provides a huge changeset that qualifies for the bump of a minor version.  Primarily, this PR re-introduces server-side builds, and performs a number of optimizations.

Updates include the following:

- Web UI
  - Add a little content to the homepage
  - Add CSS styling to the pages
  - The dropdown is now a `<datalist>`, allowing for type-to-search
  - Individual tests for APIs can be tested under a specified scope
  - There is now a dedicated 404 page
- Build/Generation
  - All tests are compiled server-side once again
  - All HTML is now generated using EJS templates
  - `MANIFESTS.json` is now reduced to just the tests, and is moved to `tests.json`
  - Endpoints are generated by the server
- Test Runtime
  - All the tests are now under `/tests/` to separate it from the backend API endpoints
  - Tests are both displayed and reported to the server in one go, with a link to submit a PR to the results GitHub repo
  - Promises are no longer used for the worker and service worker; thus, core-js is no longer a dependency
- Tests
  - `api.console` is now mapped to `api.Console`
  - The `toString` operation is now added
  - Added support custom API tests for non-members (i.e. `api.AbortSignal.abort_event` or `api.BaseAudioContext.createPeriodicWave.disableNormalisation_supported`)
  - Added test for `api.DOMTokenList.remove_whitespace_and_duplicates`
- Misc.
  - Nodemon is now used to auto-reload the server when any files change

Fixes #413, fixes #414, fixes #417, fixes #418, fixes #420, fixes #437.

### CODE REVIEW

**Comment by vinyldarkscratch** [comment]
I'm requesting your review, @foolip, since this is a big change, and I want to make sure it's all good before we merge and deploy!

File: app.js [hunk 1]
```diff
@@ -64,21 +64,70 @@ function catchError(err, res) {
...
+function generateTestPage(endpoint) {
+  const theseTests = tests.getTests(endpoint);
+  let testScope = tests.getScope(endpoint);
+  let individual = false;
+
+  const lines = [
+    '<!DOCTYPE html>',
+    '<html>',
+    '<head>',
+    '<!--Copyright 2020 Google LLC',
```
Comment (foolip) [inline app.js:76]: Let's drop this. A big comment at the top is annoying when viewing source to understand the test. If I hear that we need this even for generated output then we can add it back at the bottom.

**Review by foolip — COMMENTED** [review]
This isn't my full review, I'll take a look at this next week. Very exciting change though, I think this will make things much easier to under for BCD reviewers.

File: app.js (hunk 1 above)
Comment (vinyldarkscratch) [reply app.js:76]: Okay, I'll go ahead and drop it!  I think it'll be all fine, since there's copyright notices elsewhere.

**Comment by vinyldarkscratch** [comment]
Whew!  I have now added prefix support into server-side builds, so we can get prefix support once again!  It functions in a similar manner to how the prefix testing worked before -- test each piece with each prefix until support is found, then move onto the next -- with different prefixed variations built into the manifest.  Unfortunately, this makes the manifest file very long and very large as a result -- I've separated it into two files.

Long story short, I think this is now ready for deploying!

File: static/resources/harness.js [hunk 2]
```diff
@@ -338,7 +338,11 @@
     css.rel = 'stylesheet';
     css.type = 'text/css';
     css.href = '/resources/style.css';
-    document.head.appendChild(css);
+    try {
+      document.head.appendChild(css);
+    } catch(e) {
+      // If the CSS fails to load, oh well
```
Comment (foolip) [inline static/resources/harness.js:344]: In which browser does `document.head.appendChild(css)` throw an exception? Any exception shouldn't be due to the CSS failing to load, which happens asynchronously, but more likely because `document.head` doesn't exist. I'm guessing this happened on IE<9, based on https://developer.mozilla.org/en-US/docs/Web/API/Document/head?

File: build.js [hunk 3]
```diff
@@ -728,57 +661,108 @@ function copyResources() {
         'sourceMappingURL=core.js.map'
     );
 
-    fs.writeFile(corejsPath, result, 'utf8', function(err) {
+    fs.writeFileSync(corejsPath, result, 'utf8', (err) => {
```
Comment (foolip) [inline build.js:664]: Putting this after the `await fs.readFile` would be good. Or, if this will all go away by not using core-js, a TODO would be fine.

File: build.js [hunk 4]
```diff
@@ -728,57 +661,108 @@ function copyResources() {
...
       if (err) return console.log(err);
     });
   });
-}
+};
 
-async function build(bcd, webref) {
+const buildManifest = (tests) => {
```
Comment (foolip) [inline build.js:670]: Not reviewing this in detail, but I note that it's not a very long function, so the server should be able to compute all of this information from the tests listing at runtime. (See also comment in app.js about `/tests/*` for how I think this can change.)

File: build.js [hunk 5]
```diff
@@ -728,57 +661,108 @@ function copyResources() {
...
-}
+
+  return manifest;
+};
+
+/* istanbul ignore next */
+const build = async (webref, bcd) => {
+  const IDLTests = buildIDL(webref);
+  const CSSTests = buildCSS(webref, bcd);
+  const tests = Object.assign(IDLTests, CSSTests);
```
Comment (foolip) [inline build.js:734]:
```suggestion
  const tests = Object.assign({}, IDLTests, CSSTests);
```

To avoid modifying `IDLTests` in a way that's not obvious.

File: build.js [hunk 6]
```diff
@@ -728,57 +661,108 @@ function copyResources() {
...
-  await writeManifest(manifest);
-  copyResources();
-}
+
+  return manifest;
+};
+
+/* istanbul ignore next */
+const build = async (webref, bcd) => {
+  const IDLTests = buildIDL(webref);
```
Comment (foolip) [inline build.js:732]: By passing in `bcd` here, we should have the information we need about which prefixes to test for.

File: non-standard.idl [hunk 7]
```diff
@@ -28,6 +28,34 @@ partial interface Document {
   attribute EventHandler onbeforecut;
   attribute EventHandler onbeforepaste;
   attribute EventHandler onsearch;
+
+  boolean contains();
```
Comment (foolip) [inline non-standard.idl:32]: Can you split out all updates to non-standard.idl into its own PR?

File: find-missing.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      }
+
+      features.push(...traverseFeatures(obj[i], identifier + i + '.'));
+    }
+  }
+
+  return features;
+};
+
+const findMissing = () => {
```
Comment (foolip) [inline find-missing.js:26]: Good stuff! From that list one can see a lot that can be added to non-standard.idl. But it's best to do so separately from this overhaul.

File: .eslintrc.json [hunk 9]
```diff
@@ -1,14 +1,26 @@
 {
     "extends": ["eslint:recommended", "google"],
-    "plugins": ["prettier"],
+    "plugins": ["prefer-arrow"],
```
Comment (foolip) [inline .eslintrc.json:3]: By now it may be incredibly hard to untangle, but doing the reformatting resulting from this and other linter changes separately from the backend overhaul would be great.

What I'd do: copy over the relevant changes to package.json and .eslintrc.json changes to a new branch, and reformat using `--fix` and cherry-picking over any changes that can't be made automatically.

When that changes is on main, merge in the main branch and resolve any conflicts in favor of what's on this branch. What remains should be the real changes on this branch.

**Review by foolip — COMMENTED** [review]
Submitting a partial review, I stopped half way through build.js.

File: static/resources/harness.js (hunk 2 above)
Comment (vinyldarkscratch) [reply static/resources/harness.js:344]: It was actually early versions of Chrome (like Chrome 4), which led me to believe that it may occur in other browsers as well (like IE).  I tried with `document.getElementsByTagName('head')[0]` as well, but had no luck there.

File: .eslintrc.json [hunk 10]
```diff
@@ -1,14 +1,26 @@
...
     "parser": "babel-eslint",
     "parserOptions": {
         "ecmaVersion": 8
     },
     "rules": {
         "comma-dangle": ["error", "never"],
-        "require-jsdoc": "off",
-        "no-unused-vars": ["error", {"varsIgnorePattern": "^_$"}]
+        "max-len": ["error", {
+            "code": 80,
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply .eslintrc.json:11]: I thought I had to redefine the entire rule if I was making any changes to it, but it looks like rule options will combine!

File: DESIGN.md [hunk 11]
```diff
@@ -47,6 +52,11 @@ The following JSON...
       "__base": "var canvas = document.createElement('canvas'); var gl = canvas.getContext('webgl'); var instance = gl.getExtension('ANGLE_instanced_arrays');",
       "__test": "return canvas && instance;",
       "drawArraysInstancedANGLE": "return true && instance && 'drawArraysInstancedANGLE' in instance;"
+    },
+    "Console": {
+      "__additional": {
+        "log.substitution_strings": "console.log('Hello, %s.', 'Bob')"
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply DESIGN.md:58]: I wanted to put some sort of filler test down in the document regardless of whether it actually functioned, just to show how the additional tests function, and then come back to replace it with a working one later.

File: app.js [hunk 12]
```diff
@@ -42,27 +45,26 @@ const github = require('./github')(
 
 const Tests = require('./tests');
 const tests = new Tests({
-  manifest: require('./MANIFEST.json'),
-  host: process.env.NODE_ENV === 'production' ?
-      'mdn-bcd-collector.appspot.com' :
-      `localhost:${PORT}`,
+  tests: require('./generated/tests.json'),
+  manifest: require('./generated/manifest.json'),
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply app.js:49]: Yeah, the list of tests should give the backend everything that it needs!  I'm thinking to save this for another time.

File: .eslintrc.json (hunk 9 above)
Comment (vinyldarkscratch) [reply .eslintrc.json:3]: Cherry-picked as #421!

File: unittest/unit/tests.js [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+      assert.equal(
+          theseTests.next('http://host.test/tests/css/properties'),
+          null
+      );
+    });
+  });
+
+  it('listMainEndpoints', () => {
+    assert.deepEqual(tests.listMainEndpoints(), [
+      '/api/interfaces',
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply unittest/unit/tests.js:129]: Collapsing the tests does sound nice, it ends up making the amount of tests in a single file a little too much, causing browser freezes (especially on older browsers).  I'm also not sure how to properly hook it up with our report function to submit PRs for all of the data to the results repo.  Will have to come back to this one at a later time.

[the remark this answers is not in the document]
Comment (foolip) [reply unittest/unit/tests.js]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/430.

File: build.js (hunk 6 above)
Comment (foolip) [reply build.js:712]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/431.

File: .eslintrc.json (hunk 9 above)
Comment (foolip) [reply .eslintrc.json]: Thanks!

Comment (foolip) [reply .eslintrc.json]: Done in https://github.com/foolip/mdn-bcd-collector/pull/421, thanks!

File: find-missing.js (hunk 8 above)
Comment (foolip) [reply find-missing.js]: Done in https://github.com/foolip/mdn-bcd-collector/pull/422.

File: non-standard.idl (hunk 7 above)
Comment (foolip) [reply non-standard.idl]: Done in https://github.com/foolip/mdn-bcd-collector/pull/427.

File: build.js (hunk 6 above)
Comment (foolip) [reply build.js:712]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/431.

File: app.js [hunk 14]
```diff
@@ -138,6 +141,16 @@ app.post('/api/results/export/github', (req, res) => {
       .catch(/* istanbul ignore next */ (err) => catchError(err, res));
 });
 
+app.all('/tests/*', (req, res) => {
+  const endpoint = `/${req.params['0']}`;
+
+  if (tests.listAllEndpoints().includes(endpoint)) {
```
[the remark this answers is not in the document]
Comment (foolip) [reply app.js]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/420.

File: build.js [hunk 15]
```diff
@@ -84,152 +65,130 @@ function getCustomTestAPI(name, member) {
...
+          prefix: prefix
+        });
       }
-      if (method === 'CSS.supports' || method === 'all') {
-        lines.push(`bcd.addTest("${ident}", ${JSON.stringify(
-            {property: name, scope: 'CSS.supports'}
-        )}, 'CSS');`);
+    }
+  } else {
+    for (const prefix1 of prefixesToTest) {
```
[the remark this answers is not in the document]
Comment (foolip) [reply build.js:166]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/431.

File: DESIGN.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (foolip) [reply DESIGN.md:58]: Alright. I think `api.DOMTokenList.remove_whitespace_and_duplicates` is a good candidate for a simple custom test for illustration purposes:
```js
var elm = document.createElement('b');
elm.className = ' foo bar foo ';
elm.classList.remove('bar');
return elm.className === 'foo';
```

File: static/resources/harness.js (hunk 2 above)
Comment (foolip) [reply static/resources/harness.js]: I see. Catching and ignoring the error is fine. Only the comment needs updating, something like "If inserting the link element fails" would be more accurate about where the exception is coming from. (The consequence will be the CSS won't load, though.)

File: DESIGN.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply DESIGN.md:58]: Sure thing, added!

File: build.js [hunk 16]
```diff
@@ -97,133 +75,114 @@ const getCustomTestAPI = (name, member) => {
   return test;
 };
 
+const getCustomSubtestsAPI = (name) => {
+  const subtests = {};
+
+  if (name in customTests.api) {
+    const testbase = customTests.api[name].__base || '';
+    if ('__additional' in customTests.api[name]) {
+      for (
```
Comment (foolip) [inline build.js:84]: I've filed https://github.com/foolip/mdn-bcd-collector/issues/432 to undo weird Prettier formatting like this.

File: static/resources/harness.js [hunk 17]
```diff
@@ -84,182 +102,74 @@
...
-  function runCSS(callback) {
-    var results = [];
-
+  function runCSS(callback, results) {
     var length = pending.length;
     for (var i = 0; i < length; i++) {
-      updateStatus('Testing ' + pending[i].name);
-      results.push(test(pending[i]));
+      if (pending[i].scope == 'CSS') {
+        updateStatus('Testing ' + pending[i].name);
```
Comment (foolip) [inline static/resources/harness.js:152]: This isn't new, but I'm nervous about doing really anything at all while the tests are running. Let's come back to this when moving more of the code generation to the server side.

File: static/resources/harness.js [hunk 18]
```diff
@@ -84,182 +102,74 @@
...
     }
 
-    pending = [];
-
     callback(results);
   }
 
-  function runWorker(callback) {
-    var results = [];
+  function runWorker(callback, results) {
```
Comment (foolip) [inline static/resources/harness.js:172]: Not in this PR, but the setup for worker tests I'm imaging when most of the of work is done on the server side is that a single large script is run in a worker, and then the results are all posted back in one go. In the window context there wouldn't be any need to keep track of pending tests.

File: static/resources/harness.js [hunk 19]
```diff
@@ -360,81 +271,93 @@
...
+  function runAndDisplay() {
+    run(function(results) {
+      var css = document.createElement('link');
+      css.rel = 'stylesheet';
+      css.type = 'text/css';
+      css.href = '/resources/style.css';
+      try {
+        document.head.appendChild(css);
+      } catch (e) {
+        // If the CSS fails to load, oh well
```
Comment (foolip) [inline static/resources/harness.js]:
```suggestion
        // If document.head isn't supported, no CSS will be loaded. That's OK.
```

File: tests.js [hunk 20]
```diff
@@ -16,45 +16,167 @@
...
 
-      if (afterIndex === -1) {
-        throw new Error(`${after} not found in test manifest`);
+    for (const [ident, test] of Object.entries(this.tests)) {
+      for (const scope of test.scope) {
+        let endpoint = '';
+        let httpsOnly = false;
+        switch (scope) {
+          case 'Window':
+            endpoint = '/api/interfaces';
```
Comment (foolip) [inline tests.js:34]: In a follow-up, maybe `/api/?scope=Window`, modulo naming of "scope" discussed elsewhere. Not only interfaces are being tested any longer, namespaces and attributes/functions on the global scope are too.

File: tests.js [hunk 21]
```diff
@@ -16,45 +16,167 @@
...
+            break;
+          case 'Worker':
```
Comment (foolip) [inline tests.js:36]: Can both 'Worker' and 'DedicatedWorker' happen? It seems like it should be one or the other.

File: tests.js [hunk 22]
```diff
@@ -16,45 +16,167 @@
...
+
+        if (endpoint) {
+          if (!(endpoint in endpoints)) {
+            endpoints[endpoint] = {
+              scope: scope,
+              httpsOnly: httpsOnly,
+              entries: []
+            };
+          }
+          if (!(ident in endpoints[endpoint].entries)) {
```
Comment (foolip) [inline tests.js:57]: This won't work. `x in y` where `y` is an array will only be true if `x` is an array index (string or number) or if it's some Array builtin like `length` or `push`. Not the cases you want to check for, I presume.

File: tests.js [hunk 23]
```diff
@@ -16,45 +16,167 @@
...
 
-      begin = afterIndex + 1;
-      if (limit) {
-        end = afterIndex + 1 + limit;
+    return endpoints;
+  }
+
+  listMainEndpoints(urlPrefix = '') {
+    return Object.keys(this.endpoints).map((item) => (
+      ['', `${urlPrefix}${item}`]
```
Comment (foolip) [inline tests.js:69]: What does the empty string mean here? I think it's a marker for "run this test by default", is that right? Being more explicit about this would be good.

File: tests.js [hunk 24]
```diff
@@ -16,45 +16,167 @@
...
+    ];
+  }
+
+  next(after) {
+    const afterURL = new URL(after);
+    if (!this.httpOnly && afterURL.protocol === 'http:') {
+      return `https://${this.host}${afterURL.pathname}`;
+    }
+    const endpoints = this.listMainEndpoints('/tests');
+    const index = endpoints.findIndex((item) => {
```
Comment (foolip) [inline tests.js:92]: What if `index` is -1 (before we add 1)? Isn't that an error case that we should do something about? As it is, will we end up running the same test in an endless cycle?

File: tests.js [hunk 25]
```diff
@@ -16,45 +16,167 @@
...
+    if (index >= endpoints.length) {
+      return null;
+    }
+
+    const endpoint = endpoints[index][1];
+
+    if (this.endpoints[endpoint.replace('/tests', '')].httpsOnly) {
+      const newUrl = `https://${this.host}${endpoint}`;
+      if (this.httpOnly) {
+        // Skip this endpoint and go to the next
```
Comment (foolip) [inline tests.js:105]: I've filed https://github.com/foolip/mdn-bcd-collector/issues/435 to perhaps simplify this later. What I did was a bit too clever IMHO, and continuing to maintain the complexity like here is unfortunate.

File: tests.js [hunk 26]
```diff
@@ -16,45 +16,167 @@
...
+        return this.next(newUrl);
       }
+      return newUrl;
+    }
+
+    return `http://${this.host}${endpoint}`;
+  }
+
+  getTests(endpoint) {
+    let idents = [];
```
Comment (foolip) [inline tests.js:115]:
```suggestion
    let idents;
```

It's assigned in both branches, let's be explicit that this empty array is never used.

File: tests.js [hunk 27]
```diff
@@ -16,45 +16,167 @@
...
+      '</head>',
+      '<body>',
+      '<p id="status">Running tests...</p>',
+      '<script>'
+    ];
+
+    for (const [ident, test] of Object.entries(theseTests)) {
+      for (const scope of test.scope) {
+        if (!testScope || scope == testScope) {
+          lines.push(`bcd.addTest("${ident}", ${JSON.stringify(test.tests)}, "${scope}");`);
```
Comment (foolip) [inline tests.js:171]: This is a place where I expect resolving https://github.com/foolip/mdn-bcd-collector/issues/376 will change things a lot. Rather than feeding input to the client-side test generator here, we should just output stuff like:
```js
bcd.addTest('api.Document', function() {
  return 'Document' in window;
});
```

But not in this PR :)

**Review by foolip — COMMENTED** [review]
Finally got through the whole change. Lots of comments, please enjoy :)

Let's punt as much as possible to issues where things aren't regressed by this change.

File: static/resources/harness.js [hunk 28]
```diff
@@ -84,182 +102,74 @@
...
-              result.result = true;
-            } else {
-              result.result = null;
-            }
+    for (var i = 0; i < data.tests.length; i++) {
+      var test = data.tests[i];
 
-            result.message = 'threw ' + stringify(err);
+      try {
+        var value = eval(test.code);
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply static/resources/harness.js:110]: The reason for `eval` is to be able to send the functions to the workers -- all messages sent to workers must be serializable, so passing in functions wouldn't work, unfortunately.

File: static/resources/harness.js [hunk 29]
```diff
@@ -84,182 +102,74 @@
...
+        if (value && typeof value === 'object' && 'result' in value) {
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply static/resources/harness.js:111]: This is a part of the addition of `bcd.testConstructor`, which returns an object (that can include a message).

File: tests.js (hunk 21 above)
Comment (vinyldarkscratch) [reply tests.js:36]: Some interfaces mark exposure for `Worker`, while others for `DedicatedWorker`.  From what I've gathered, they're both the same however.

File: package.json [hunk 30]
```diff
@@ -38,11 +38,12 @@
     "json-stable-stringify": "1.0.1",
     "json3": "3.3.3",
     "mdn-browser-compat-data": "1.0.35",
+    "nodemon": "^2.0.4",
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply package.json:41]: Overall it shouldn't interfere with GAE, but just in case, I'll split this into `start-dev` instead.  Nodemon watches for any file changes and automatically reloads the server accordingly -- SUPER helpful for development purposes.

File: static/index.html [hunk 31]
```diff
@@ -1,69 +1,126 @@
...
+    function getTests() {
+      var button = document.getElementById('start');
+      var dropdown = document.getElementById('tests');
+      var individualTestOptionsBox = document.getElementById('individualTestOptionsBox');
+      var secureContext = document.getElementById('secureContext');
+      var limitScope = document.getElementById('limitScope');
 
-<p><a href="https://github.com/foolip/mdn-bcd-collector">Source on GitHub</a></p>
-</body>
+      var client = new XMLHttpRequest();
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply static/index.html]: Since it's within the scope of server-side builds, I decided to actually tackle this, and hook up EJS to build the pages.  All pages -- including the homepage, results, and the tests themselves -- are now built with the EJS templating system!

File: static/resources/harness.js (hunk 17 above)
Comment (vinyldarkscratch) [reply static/resources/harness.js:152]: The `updateStatus()` isn't really working due to how much is going on in the JavaScript already (the DOM isn't updating), so these can be taken out safely.

File: tests.js (hunk 20 above)
Comment (vinyldarkscratch) [reply tests.js:34]: I just pushed f8934be379e465cdc78a2d514d0f3a6d6c164380 which should provide some assistance for this.  In that commit, I added `?reportToServer=true`, which will tell the website to send the data to the server rather than display it.  It no longer counts on a test being a "main endpoint" to determine if it should report, which works towards eliminating main endpoints altogether.

File: views/index.ejs [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+    } else {
+      limitExposureBox.style.display = "block";
+    }
+  }
+
+  button.onclick = function() {
+    var choice = getDropdownSelection(dropdown);
+    var newUrl = choice[1];
+    if (choice[0] > 0) {
+      var protocol = secureContext.checked ? 'https:' : 'http:';
```
Comment (foolip) [inline views/index.ejs:71]: Could this use the protocol of the current page instead? For localhost 'https:' will never make sense, so showing the option then isn't really useful.

File: views/layout.ejs [hunk 33]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <!--Copyright 2020 Google LLC
```
Comment (foolip) [inline views/layout.ejs:4]: Let's omit copyright notices from the generated pages. I shouldn't have included it in the first place :)

File: views/tests.ejs [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+          background-color: #111;
+          color: white;
+        }
+      }
+    </style>
+  </head>
+  <body>
+    <p id="status">Running tests...</p>
+    
+    <script src="/resources/json3.min.js"></script>
```
Comment (foolip) [inline views/tests.ejs:28]: Can you add a TODO linking to https://github.com/foolip/mdn-bcd-collector/issues/407 here?

**Review by foolip — COMMENTED** [review]
All done reviewing, phew!

I think this looks pretty good now, I'd like to get it landed so we can keep iterating in smaller PRs. I've tried deploying https://staging-dot-mdn-bcd-collector.appspot.com/ but something is broken.

Once we have a working staging deployment, it would be good to compare the collected results before/after this change, to see if there are unintended changes in there.

**Comment by foolip** [comment]
It looks like the 500 error on https://staging-dot-mdn-bcd-collector.appspot.com/ is because the 256 MB limit is being exceeded. @vinyldarkscratch does that give any clues? Feel free to push changes to the staging branch to narrow down what it might be that causes this.

File: nodemon.json [hunk 35]
```diff
@@ -1,4 +1,9 @@
 {
   "verbose": true,
-  "ignore": ["static/*", "coverage/*", ".nyc_output/*"],
+  "ignore": [
+    "static/*",
+    "views/*",
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply nodemon.json:5]: It's because the server itself doesn't need to reload for that -- just the browser.

File: tests.js (hunk 24 above)
Comment (vinyldarkscratch) [reply tests.js:93]: Fixed down below!  If the `index` is 0 (`-1 + 1`), we return `null` now, to prevent an endless loop.

File: build.js [hunk 36]
```diff
@@ -624,12 +669,12 @@ const buildCSS = (webref, bcd) => {
...
         code: getCustomTestCSS(name) || [
-          {property: attrName, scope: 'document.body.style'},
-          {property: name, scope: 'CSS.supports'}
+          {property: attrName, owner: 'document.body.style'},
+          {property: name, owner: 'CSS.supports'}
         ],
         combinator: '||'
       },
-      scope: ['CSS']
+      exposure: ['CSS']
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply build.js]: Yeah, `exposure` was certainly misused here.  I'm now adding a `category` property, which will also help future-proof for `javascript` feature testing or other categories.

File: static/resources/harness.js [hunk 37]
```diff
@@ -140,16 +140,15 @@
...
 
     return result;
   }
 
   function runCSS(callback, results) {
     var length = pending.length;
     for (var i = 0; i < length; i++) {
-      if (pending[i].scope == 'CSS') {
-        updateStatus('Testing ' + pending[i].name);
+      if (pending[i].exposure == 'CSS') {
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply static/resources/harness.js]: I've now fixed up `exposure` and removed this `runCSS()` function entirely!

File: build.js [hunk 38]
```diff
@@ -624,12 +669,12 @@ const buildCSS = (webref, bcd) => {
     tests[`css.properties.${name}`] = compileTest({
       raw: {
         code: getCustomTestCSS(name) || [
-          {property: attrName, scope: 'document.body.style'},
-          {property: name, scope: 'CSS.supports'}
+          {property: attrName, owner: 'document.body.style'},
+          {property: name, owner: 'CSS.supports'}
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply build.js:680]: `CSS.supports` is kind of a special case in this sense, as it's the only instance of "how do we test it".  We could, perhaps, call it "method" in this case?

**Comment by vinyldarkscratch** [comment]
> It looks like the 500 error on https://staging-dot-mdn-bcd-collector.appspot.com/ is because the 256 MB limit is being exceeded. @vinyldarkscratch does that give any clues? Feel free to push changes to the staging branch to narrow down what it might be that causes this.

What is the 256 MB limit for?  Is it storage space, or how large the response from the server is?  (Unfortunately my knowledge of GAE is limited...)

**Comment by vinyldarkscratch** [comment]
Great news: I was able to eliminate "main endpoints" in an effective manner, so we now only have endpoints for the individual tests!  My solution was to have the script both submit AND display the results, no redirects, with a link at the top to submit the results to the GitHub repo.

**Comment by foolip** [comment]
@vinyldarkscratch other than this not working on AppEngine, is this ready for final review, or do you have more changes planned?

**Comment by vinyldarkscratch** [comment]
I'd say this is ready for final review!

**Comment by vinyldarkscratch** [comment]
I want to try to integrate the datalist if possible, and simplify the code some if I can.  I should be ready for a final review soon, though!

**Comment by vinyldarkscratch** [comment]
Okay, this is now ready for final review, @foolip!

File: build.js (hunk 38 above)
[the remark this answers is not in the document]
Comment (foolip) [reply build.js:621]: That would be fine, yeah. I won't block this PR on fixing this though, I've linked it in https://github.com/foolip/mdn-bcd-collector/issues/375 to not forget.

File: unittest/app/app.js [hunk 39]
```diff
@@ -109,4 +106,68 @@ describe('/api/results', () => {
...
+    const res = await agent.get(`/tests/${tests[1][0].replace(/\./g, '/')}`);
+    assert.equal(res.status, 200);
+  });
+
+  it('get all tests', async () => {
+    const res = await agent.get('/tests/');
+    assert.equal(res.status, 200);
+  });
+
+  it('get a non-existent tests', async () => {
```
Comment (foolip) [inline unittest/app/app.js:139]: Good testing!

**Review by foolip — APPROVED** [review]
Let's get this merged!!!
